File: src/main/java/com/extendedae_plus/integration/jade/CommonProviders.java
package com.extendedae_plus.integration.jade;

import appeng.api.networking.IGridConnection;
import appeng.me.helpers.IGridConnectedBlockEntity;
import com.extendedae_plus.common.registry.block.wirelessTransceiver.BlockEntityWirelessTransceiver;
import com.extendedae_plus.common.wireless.linkApi.IBlockEntityLabel;
import com.extendedae_plus.common.wireless.linkApi.Label;
import com.extendedae_plus.common.wireless.linkApi.RegistryLink;
import com.extendedae_plus.integration.IntegrationFTBTeams;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import org.jetbrains.annotations.Nullable;
import snownee.jade.api.BlockAccessor;

import java.util.UUID;
import java.util.function.BiConsumer;
import java.util.function.Function;

public final class CommonProviders {
    public static final BiConsumer<CompoundTag, BlockAccessor> linkLabel = (data, accessor) -> {
        if (!(accessor.getBlockEntity() instanceof IBlockEntityLabel blockEntity)) return;

        var label = blockEntity.getLabel().data;
        if (label.isEmpty()) {
            data.putBoolean("unset", true);
        } else if (label.frequency() != null) {
            data.putLong("frequency", label.frequency());
        } else data.putString("label", label.label());
    };

    public static final BiConsumer<CompoundTag, BlockAccessor> linkChannels = (data, accessor) -> {
        if (!(accessor.getBlockEntity() instanceof IGridConnectedBlockEntity blockEntity)) return;
        var node = blockEntity.getGridNode();
        if (node == null) return;

        int usedChannels = 0;
        for (IGridConnection connection : node.getConnections()) {
            usedChannels = Math.max(connection.getUsedChannels(), usedChannels);
        }
        data.putInt("used", usedChannels);
        data.putInt("max", node.getMaxChannels());
    };

    public static BiConsumer<CompoundTag, BlockAccessor> locationMaster(Function<BlockAccessor, Label> infoExternal) {
        return (data, accessor) -> {
            var label = infoExternal.apply(accessor);
            if (label == null) return;

            var masterHost = RegistryLink.findMaster(label);
            if (masterHost == null || masterHost.isRemoved()) return;

            var pos = masterHost.getBlockPos();
            var level = masterHost.getServerLevel();
            if (pos != null)
                data.putLong("pos", masterHost.getBlockPos().asLong());
            if (level != null)
                data.putString("dim", level.dimension().location().toString());
            if (pos != null && level != null) {
                if (level.getBlockEntity(pos) instanceof BlockEntityWirelessTransceiver masterBlockEntity
                        && masterBlockEntity.hasCustomName())
                    data.putString("name", masterBlockEntity.getCustomName().getString());
            }
        };
    }

    public static BiConsumer<CompoundTag, BlockAccessor> stateLocked(BooleanProperty propertyLocked) {
        return (data, accessor) -> {
            var blockState = accessor.getBlockState();
            var locked = blockState.getOptionalValue(propertyLocked);
            if (locked.isEmpty()) return;
            data.putBoolean("locked", locked.get());
        };
    }

    public static BiConsumer<CompoundTag, BlockAccessor> infoPlacer(Function<BlockAccessor, @Nullable UUID> infoExternal) {
        return (data, accessor) -> {
            var placer = infoExternal.apply(accessor);
            if (placer == null) return;
            data.putUUID("uuid", placer);
            data.putString("name", IntegrationFTBTeams.instance.getTeamName(placer).getString());
        };
    }
}

File: src/main/kotlin/com/extendedae_plus/integration/impl/jade/CommonProviders.kt
package com.extendedae_plus.integration.impl.jade

import appeng.me.helpers.IGridConnectedBlockEntity
import com.extendedae_plus.common.registry.block.wirelessTransceiver.TileWirelessTransceiver
import com.extendedae_plus.common.wireless.linkApi.IBlockEntityLabel
import com.extendedae_plus.common.wireless.linkApi.Label
import com.extendedae_plus.common.wireless.linkApi.RegistryLink
import com.extendedae_plus.integration.helper.ManagerIntegration
import com.extendedae_plus.integration.impl.point.IntegrationFTBTeams
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.block.state.properties.Property
import snownee.jade.api.BlockAccessor
import java.util.*
import kotlin.jvm.optionals.getOrNull

private typealias Function<A, B> = (A) -> B

object CommonProviders {
    @JvmField
    val linkLabel = provider@{ data: CompoundTag, accessor: BlockAccessor ->
        val tile = accessor.blockEntity as? IBlockEntityLabel ?: return@provider
        val label = tile.label.data
        if (label.isEmpty) {
            data.putBoolean("unset", true)
        } else if (label.frequency != null) {
            data.putLong("frequency", label.frequency)
        } else data.putString("label", label.label)
    }

    @JvmField
    val linkChannels = provider@{ data: CompoundTag, accessor: BlockAccessor ->
        val node = (accessor.blockEntity as? IGridConnectedBlockEntity)?.gridNode ?: return@provider

        var usedChannels = 0
        for (connection in node.connections) {
            usedChannels = connection.usedChannels.coerceAtLeast(usedChannels)
        }
        data.putInt("used", usedChannels)
        data.putInt("max", node.maxChannels)
    }

    @JvmStatic
    fun locationMaster(infoExternal: Function<BlockAccessor, Label?>) =
        provider@{ data: CompoundTag, accessor: BlockAccessor ->
            val label = infoExternal(accessor) ?: return@provider

            val masterHost = RegistryLink.findMaster(label)
            if (masterHost == null || masterHost.isRemoved) return@provider

            val pos = masterHost.blockPos
            val level = masterHost.serverLevel
            data.putLong("pos", masterHost.blockPos.asLong())
            if (level != null) data.putString("dim", level.dimension().location().toString())
            if (level != null) {
                val blockEntity = level.getBlockEntity(pos)
                        as? TileWirelessTransceiver
                    ?: return@provider
                if (blockEntity.hasCustomName())
                    data.putString("name", blockEntity.customName?.string ?: return@provider)
            }
        }

    @JvmStatic
    fun stateLocked(propertyLocked: Property<Boolean>) = provider@{ data: CompoundTag, accessor: BlockAccessor ->
        val blockState = accessor.blockState
        val locked = blockState
            .getOptionalValue(propertyLocked)
            .getOrNull()
            ?: return@provider
        data.putBoolean("locked", locked)
    }

    @JvmStatic
    fun infoPlacer(infoExternal: Function<BlockAccessor, UUID?>) =
        provider@{ data: CompoundTag, accessor: BlockAccessor ->
            val placer = infoExternal(accessor) ?: return@provider
            data.putUUID("uuid", placer)
            data.putString("name", ManagerIntegration<IntegrationFTBTeams>()
                ?.getTeamName(placer)
                ?.string ?: return@provider)
        }
}
